lib/OpenQA/WebAPI/Controller/LiveViewHandler.pm

```diff
@@ -0,0 +1,251 @@
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+package OpenQA::WebAPI::Controller::LiveViewHandler;
+use strict;
+use Try::Tiny;
+use Mojo::URL;
+use Mojo::Base 'OpenQA::WebAPI::Controller::Developer';
+use OpenQA::Utils;
+use OpenQA::Schema::Result::Jobs;
+use JSON qw(decode_json);
+
+# notes: * routes in this package are served by LiveViewHandler rather than the regular WebAPI server
+#        * using prefork is currently not possible
+
+# define a whitelist of commands to be passed to os-autoinst via ws_proxy
+my %allowed_os_autoinst_commands = (
+    set_pause_at_test     => 1,
+    resume_test_execution => 1,
+);
+
+# define global variable to store ws connections (tx) to isotovideo cmd srv for each job
+my %cmd_srv_transactions_by_job;
+
+# define global variable to store ws connections (tx) to JavaScript clients for each job
+my %java_script_transactions_by_job;
+
+# provides a web socket connection acting as a proxy to interact with os-autoinst indirectly
+sub ws_proxy {
+    my ($self) = @_;
+
+    # determine basic variables
+    my $job  = $self->find_current_job() or return $self->reply->not_found;
+    my $user = $self->current_user()     or return $self->reply->not_found;
+    my $app  = $self->app;
+    my $job_id             = $job->id;
+    my $user_id            = $user->id;
+    my $developer_sessions = $app->schema->resultset('DeveloperSessions');
+
+    $app->log->debug('ws_proxy: client connected: ' . $user->name);
+
+    # define variables for transactions
+    my $java_script_tx                           = $self->tx;
+    my $java_script_tx_id                        = "$job_id-$java_script_tx";
+    my $java_script_transactions_for_current_job = ($java_script_transactions_by_job{$job_id} //= []);
+    push(@$java_script_transactions_for_current_job, $java_script_tx);
+
+    # register development session, ensure only one development session is opened per job
+    my $developer_session = $developer_sessions->register($job_id, $user_id);
+    if (!$developer_session) {
+        return $self->render(
+            json => {
+                error => 'unable to create (further) development session'
+            },
+            status => 400
+        );
+    }
+
+    # mark session as active
+    $developer_session->update({ws_connection_count => \'ws_connection_count + 1'});    #'restore syntax highlighting
+
+    # define functions to push information to the JavaScript client
+    my $send_message_to_java_script = sub {
```

Comment:
i would move those inside the object

Reply:
Making those lambdas has the advantage of having the context variables directly available. So there's less parameter passing, better encapsulation and the code is less cluttered. But I can play around and see how it would look like when moving those to the controller object.

Reply:
Yeah, but it's more difficult to test it then and less readable imho. So the encapsulation advantage gets kinda lost, it's not only about style (parameter passing / rely on context).  But this likely will bite you when writing unit tests, as it will force lot of hacks to hit all the codepath.

From a personal perspective i prefer even more parameter passing than relying on context, as it tends to grow and it's more difficult to read, up to you.

Reply:
> But this likely will bite you when writing unit tests, as it will force lot of hacks to hit all the codepath.

That might justify the effort to refactor this, indeed.